lib/charms/observability_libs/v1/cert_handler.py

```diff
@@ -87,15 +89,164 @@ class CertHandlerEvents(ObjectEvents):
     cert_changed = EventSource(CertChanged)
 
 
+class _VaultBackend(abc.ABC):
+    def store(self, contents: Dict[str, str], clear: bool = False): ...
+
+    def get_value(self, key: str) -> Optional[str]: ...
+
+    def retrieve(self) -> Dict[str, str]: ...
+
+    def nuke(self): ...
+
+
+class _RelationVaultBackend(_VaultBackend):
+    """Relation backend for Vault.
+
+    Use it to store data in a relation databag.
+    Assumes that a single relation exists and its data is readable.
+    If not, it will raise RuntimeErrors as soon as you try to read/write.
+    It will store the data, in plaintext (json-dumped) nested under a configurable
+    key in the **unit databag** of this relation.
+    """
+
+    def __init__(self, charm: CharmBase, relation_name: str, nest_under: str = "secret-contents"):
+        self.charm = charm
+        self.relation_name = relation_name
+        self.nest_under = nest_under  # needs to be charm-unique.
+
+    def _check_ready(self):
+        try:
+            self.charm.model.get_relation(self.relation_name).data[  # pyright: ignore
+                self.charm.unit
+            ]
+        except Exception as e:
+            # if something goes wrong here, the peer-backed vault is not ready to operate
+            # it can be because you are trying to use it too soon, i.e. before the peer
+            # relation has been created (or has joined).
+            raise RuntimeError(" backend not ready.") from e
+
+    @property
+    def _relation(self) -> Optional[Relation]:
+        self._check_ready()
+        return self.charm.model.get_relation(self.relation_name)
+
+    @property
+    def _databag(self):
+        self._check_ready()
+        return self._relation.data[self.charm.unit]  # pyright: ignore
+
+    def _read(self) -> Dict[str, str]:
+        value = self._databag.get(self.nest_under)
+        if value:
+            return json.loads(value)
+        return {}
+
+    def _write(self, value: Dict[str, str]):
+        if not all(isinstance(x, str) for x in value.values()):
+            # the caller has to take care of encoding
+            raise TypeError("You can only store strings in Vault.")
+
+        self._databag[self.nest_under] = json.dumps(value)
+
+    def store(self, contents: Dict[str, str], clear: bool = False):
+        """Create a new revision by updating the previous one with ``contents``."""
+        current = self._read()
+
+        if clear:
+            current.clear()
+
+        current.update(contents)
+        self._write(current)
+
+    def get_value(self, key: str):
+        """Like retrieve, but single-value."""
+        return self._read().get(key)
+
+    def retrieve(self):
+        """Return the full vault content."""
+        return self._read()
+
+    def nuke(self):
```

Comment:
Surprisingly, `woke` is offended by `dummy` or `sanity` more than by `nuke`.
```suggestion
    def clear(self):
```